File: solutions/098.rb
#  <p>By replacing each of the letters in the word CARE with $1$, $2$, $9$, and $6$
#  respectively, we form a square number: $1296 = 36^2$. What is remarkable is tha
# t, by using the same digital substitutions, the anagram, RACE, also forms a squa
# re number: $9216 = 96^2$. We shall call CARE (and RACE) a square anagram word pa
# ir and specify further that leading zeroes are not permitted, neither may a diff
# erent letter have the same digital value as another letter.</p>
# <p>Using <a href="resources/documents/0098_words.txt">words.txt</a> (right click
#  and 'Save Link/Target As...'), a 16K text file containing nearly two-thousand c
# ommon English words, find all the square anagram word pairs (a palindromic word
# is NOT considered to be an anagram of itself).</p>
# <p>What is the largest square number formed by any member of such a pair?</p>
# <p class="smaller">NOTE: All anagrams formed must be contained in the given text
#  file.</p>

require 'set'

# Read words.txt
# Assuming words.txt is in the same directory as the script.
# The problem statement link is "resources/documents/0098_words.txt".
# Using 'words.txt' as per the original script.
words = File.read('words.txt').gsub(/"/, '').split(',').map(&:strip)

# Determine maximum word length
max_len = 0
words.each { |w| max_len = [max_len, w.length].max }

# Precompute squares, grouped by length
# No filter for unique digits in the square itself here.
squares_by_length = Hash.new { |h, k| h[k] = Set.new }
n = 1
loop do
  square = n * n
  s_str = square.to_s
  break if s_str.length > max_len
  squares_by_length[s_str.length].add(square)
  n += 1
end

# Group words by sorted letters (signature) to find anagrams
# Only consider groups that can actually form pairs
anagram_groups = words.group_by { |w| w.chars.sort.join }.values.select { |group| group.length > 1 }

max_found_square = 0

anagram_groups.each do |group|
  word_len = group[0].length # All words in an anagram group have the same length
  candidate_squares = squares_by_length[word_len]
  next if candidate_squares.empty?

  group.combination(2).each do |word1, word2|
    # word1 and word2 are anagrams and have length word_len

    candidate_squares.each do |square1_val|
      s1_chars = square1_val.to_s.chars # Length is already word_len

      mapping = {}
      possible_map = true
      
      # Build mapping from word1 to s1_chars
      word1.chars.each_with_index do |char, i|
        digit = s1_chars[i]
        if mapping.key?(char)
          # If char already mapped, ensure consistency
          if mapping[char] != digit
            possible_map = false
            break
          end
        else
          # New char to map.
          # The check for "different letter have same digital value"
          # is done after forming the full mapping.
          mapping[char] = digit
        end
      end
      next unless possible_map

      # Check: "neither may a different letter have the same digital value as another letter."
      # This means the number of unique digits used must equal the number of unique letters in the word.
      # mapping.keys contains unique letters from word1.
      # mapping.values contains the digits they map to.
      # If word1 is "NOON", mapping.keys=["N","O"]. mapping.values might be ["1","0"]. Uniq count is 2. OK.
      # If mapping results in N->1, O->1, then mapping.values=["1","1"]. Uniq count is 1. Not OK.
      next if mapping.values.uniq.length != mapping.keys.length


      # Form number string for word2 using the mapping
      s2_chars = word2.chars.map { |char_in_w2| mapping[char_in_w2] }
      # Anagrams have same set of characters, so all char_in_w2 will be in mapping.keys
      s2_str = s2_chars.join

      # Validate s2_str
      # 1. Must have correct length (guaranteed by map from word2 of same length)
      #    but check anyway for safety, though map should handle it.
      next if s2_str.length != word_len
      
      # 2. No leading zero (if length > 1)
      next if word_len > 1 && s2_str[0] == '0'

      square2_val = s2_str.to_i

      # Check if square2_val is a valid square of the correct length
      if candidate_squares.include?(square2_val)
        # Found a square anagram pair (word1, square1_val) and (word2, square2_val)
        max_found_square = [max_found_square, square1_val, square2_val].max
      end
    end
  end
end

puts "Largest square number: #{max_found_square}"

File: python/098.py
#!/usr/bin/env python3
"""
Anagramic squares (Problem 98)

Find the largest square number formed by any member of a square anagram word pair.
"""

import math
from typing import Dict, List, Set


def main() -> None:
    """Find largest square anagram pair."""
    from pathlib import Path
    script_dir = Path(__file__).parent
    data_file = script_dir.parent / 'data' / 'words.txt'
    
    # Read words
    with open(data_file) as f:
        words = [w.strip().strip('"') for w in f.read().split(',')]
    
    # Determine maximum word length
    max_len = max(len(w) for w in words)
    
    # Precompute squares grouped by length
    squares_by_length: Dict[int, Set[int]] = {}
    n = 1
    while True:
        square = n * n
        s_str = str(square)
        if len(s_str) > max_len:
            break
        length = len(s_str)
        if length not in squares_by_length:
            squares_by_length[length] = set()
        squares_by_length[length].add(square)
        n += 1
    
    # Group words by sorted letters (signature) to find anagrams
    anagram_groups: Dict[str, List[str]] = {}
    for word in words:
        signature = ''.join(sorted(word))
        if signature not in anagram_groups:
            anagram_groups[signature] = []
        anagram_groups[signature].append(word)
    
    # Filter to groups with pairs
    anagram_groups = {k: v for k, v in anagram_groups.items() if len(v) > 1}
    
    max_found_square = 0
    
    for group in anagram_groups.values():
        word_len = len(group[0])
        candidate_squares = squares_by_length.get(word_len, set())
        if not candidate_squares:
            continue
        
        # Check all pairs in the group
        for i in range(len(group)):
            for j in range(i + 1, len(group)):
                word1, word2 = group[i], group[j]
                
                for square1_val in candidate_squares:
                    s1_chars = list(str(square1_val))
                    
                    # Build mapping from word1 to square1
                    mapping: Dict[str, str] = {}
                    possible_map = True
                    
                    for char_idx, char in enumerate(word1):
                        digit = s1_chars[char_idx]
                        if char in mapping:
                            if mapping[char] != digit:
                                possible_map = False
                                break
                        else:
                            mapping[char] = digit
                    
                    if not possible_map:
                        continue
                    
                    # Check: different letters must map to different digits
                    if len(set(mapping.values())) != len(mapping):
                        continue
                    
                    # Form number for word2 using mapping
                    s2_chars = [mapping[char] for char in word2]
                    s2_str = ''.join(s2_chars)
                    
                    if len(s2_str) != word_len:
                        continue
                    
                    if word_len > 1 and s2_str[0] == '0':
                        continue
                    
                    square2_val = int(s2_str)
                    
                    if square2_val in candidate_squares:
                        max_found_square = max(max_found_square, square1_val, square2_val)
    
    print(max_found_square)


if __name__ == "__main__":
    main()
